File: Java/题目1522：包含min函数的栈.java

// 题目1522：包含min函数的栈

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1522
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17242701
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Stack;

public class Main {
    /*
     * 1522
     */
public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            int n = (int) st.nval;
            Stack<Integer> stack1 = new Stack<Integer>();
            Stack<Integer> stack2 = new Stack<Integer>();
            for (int i = 0; i < n; i++) {
            	st.nextToken();
                String a = st.sval;
                if (a.equals("o")) {
                    if (stack1.isEmpty()) {
                        System.out.println("NULL");
                    }else {
                        stack1.pop();
                        stack2.pop();
                        if (stack2.isEmpty()) {
                            System.out.println("NULL");
                        }else {
                            System.out.println(stack2.peek());
                        }
                    }
                }else if (a.contains("s")) {
                	st.nextToken();
                	int tempNum = (int) st.nval;
                    stack1.push(tempNum);
                    if (stack2.isEmpty()) {
                        stack2.push(tempNum);
                    }else {
                        if (tempNum < stack2.peek()) {
                            stack2.push(tempNum);
                        }else {
                            stack2.push(stack2.peek());
                        }
                    }
                    System.out.println(stack2.peek());
                }
            }
        }
    }}
/**************************************************************
 Problem: 1522
 User: wangzhenqing
 Language: Java
 Result: Accepted
 Time:880 ms
 Memory:27384 kb
 ****************************************************************/

File: C++/题目1522：包含min函数的栈.cpp

// 题目1522：包含min函数的栈

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1522
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17242701
 */

#include <stdio.h>
#include <stack>   
#include <string.h>
#include <string>

using namespace std;
int n, i;

int main() {
    while (scanf("%d", &n) != EOF) {
        stack<int> numStack;
        stack<int> minStack;
        for (i = 0; i < n; i++) {
            char operate[2];
            scanf("%s", operate);
            if (operate[0] == 'o') {
                if (numStack.empty()) {
                    printf("NULL\n");
                } else {
                    numStack.pop();
                    minStack.pop();
                    if (minStack.empty()) {
                        printf("NULL\n");
                    } else {
                        printf("%d\n", minStack.top());
                    }
                }
            } else {
                int k;
                scanf("%d", &k);
                numStack.push(k);
                if (minStack.empty()) {
                    minStack.push(k);
                } else {
                    if (k < minStack.top()) {
                        minStack.push(k);
                    } else {
                        minStack.push(minStack.top());
                    }
                }
                printf("%d\n", minStack.top());
            }
        }

    }
    return 0;
}

/**************************************************************
Problem: 1522
User: wangzhenqing
Language: C++
Result: Accepted
Time:20 ms
Memory:1052 kb
****************************************************************/


                        

